cmd, consensus/ethash, eth: miner push notifications

karalabe opened the pull request:
This PR introduces a new flag `--miner.notify`, which accepts a comma-separated list of URLs to push miner work packages to. It's main use case is for mining pools to be pinged by new blocks as fast as possible, without requiring to constantly poll the `miner.getWork` endpoint.

Supersedes https://github.com/ethereum/go-ethereum/pull/17324.

karalabe commented:
@ppratscher @rjl493456442 PTAL

rjl493456442 commented on consensus/ethash/sealer.go:
```diff
@@ -161,40 +164,65 @@ search:
-	// getWork returns a work package for external miner.
+		for i, url := range notify {
+			// Terminate any previously pending request and create the new work
+			if notifyReqs[i] != nil {
```
Should we set the `notifyReq[i]` as nil if the request has been sent successfully?

karalabe replied:
That would be racey. To be fair however, this current code might be racey too. I'll think about it.

ppratscher commented:
LGTM!

rjl493456442 commented:
Just one comment, otherwise LGTM

karalabe commented:
@rjl493456442 I fixed the data race by passing `notifyReqs[i]` as a parameter to the closure instead of only the index. This way the `remote` goroutine retains ownership over `notifyReqs`. I've also added a tests that pushes a lot of work packets fast. Previously the race detector was whining, now it's happy. PTAL.
